torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+        else:
+            raise ValueError(f"Unknown StateDictType {self.state_dict_type}.")
+        return state_dict
+
+    def state_dict(self, destination=None, prefix="", keep_vars=False):
+        """
+        The entry point of all three FSDP state_dict APIs.
+        ``self.state_dict_type`` decides which code path to execute.
+
+        .. warning:: This needs to be called on all ranks, since synchronization
+            primitives may be used.
+        """
+        torch.cuda.synchronize()
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            state_dict = self.module.flat_state_dict(destination, prefix, keep_vars)
+            state_dict = self._post_state_dict_hook(self, state_dict, prefix=prefix)
+        elif self.state_dict_type == StateDictType.SHARDED_STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        else:
+            raise ValueError(f"Unknown StateDictType {self.state_dict_type}.")
+        return state_dict
+
+    def local_state_dict(self, *args: Any, **kwargs: Any) -> Any:
+        """
+        Returns the local state of the module. Parameters are flattened and
+        sharded, so the resulting state_dict can only be loaded after the module
+        has been wrapped with FSDP.
+        """
+        with self.set_state_dict_type(StateDictType.LOCAL_STATE_DICT):
+            return FullyShardedDataParallel.state_dict(self, *args, **kwargs)
+
+    @staticmethod
+    def _pre_load_state_dict_hook(
+        module: nn.Module,
+        state_dict: Union[Dict[str, torch.Tensor], "OrderedDict[str, torch.Tensor]"],
+        prefix: str,
+        *args: Any,
+    ) -> None:
+        """
+        ``_pre_state_dict_hook` is called before ``self._load_from_state_dict()``
+        is called. ``self.state_dict_type`` is used to decide what preprocessing
+        will be done.
+        """
+        self = cast(FlatParameter, module)
+        if self.state_dict_type == StateDictType.STATE_DICT:
+            raise NotImplementedError("Will be implemented in the next PRs.")
+        elif self.state_dict_type == StateDictType.LOCAL_STATE_DICT:
+            _replace_by_prefix(state_dict, prefix, f"{prefix}_fsdp_wrapped_module.")
+            flat_param = self.module.flat_param
+            key = f"{prefix}_fsdp_wrapped_module.flat_param"
+            load_tensor = state_dict[key]
+            assert isinstance(
+                load_tensor, ShardedTensor
+            ), "Tensors in local_state_dict should be ShardedTensor."
+            shards = load_tensor.local_shards()
+            assert len(
+                shards
+            ), "load_local_state_dict assume one shard per ShardedTensor."
+            load_tensor = cast(torch.Tensor, shards[0].tensor)
```

Comment:
Can we separate out any of this logic into helper functions? Such as the process to reconstruct the sharded parameter from `shards` object. 

Also, same comment around adding comments/documentation around what this code is doing for better maintainability / understanding.